### PR TITLE
feat(item): counter formatter to customize counter text display

### LINKED ISSUE #24327: feat: counterFormatter for counter string in `ion-item`
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already include this feature request, without success.

### Describe the Feature Request

Introduced in v6 is the ability for developers to use `counter` with an `ion-item`, to display max length validation to the user. Developers should have complete control over the formatted appearance of this counter. Currently they are stuck with `5 / 10` formatting.

### Describe the Use Case

Users that are less technical or are unfamiliar of max length formatting, may be confused when seeing a `5 / 10` label. This feature would allow developers to customize the format of the display to say "4 characters remaining" or "5 out of 10 words".

### Describe Preferred Solution

`ion-item` exposes a formatter function that takes in the input's text content length and the max length validation and returns a formatted string.

### Describe Alternatives

_No response_

### Related Code

```ts
counterFormatter(inputLength, maxLength) {
   return `${maxLength - inputLength} characters remaining`;
}
```

```html
<ion-input [counter]="true" [counterFormatter]="counterFormatter">
   <ion-label position="stacked">Input</ion-label>
   <ion-input maxlength="20"></ion-input>
</ion-input>
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

The counter text displays as `inputLength/maxLength` with no customization available to developers.

Issue Number: #24327

## What is the new behavior?

Introduces a `counterFormatter` function on `ion-item` that allows developers to customize the display of the counter string.

The default behavior of `counterFormatter` is to display the text as `inputLength / maxLength`, such as `5 / 10` to align with the Material Design Specification.

Example use cases of this formatter could include displaying the remaining characters, such as "5 characters remaining" or giving additional context to what the max length represents, such as "5 / 10 characters". 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by amandaesmith3 — APPROVED** [review]
🚀

File: core/src/components/item/item.tsx [hunk 1]
```diff
@@ -134,8 +136,21 @@ export class Item implements ComponentInterface, AnchorInterface, ButtonInterfac
...
+   * A callback used to format the counter text.
+   * By default the counter text is set to "itemLength / maxLength".
+   */
+  @Prop() counterFormatter: CounterFormatter = (itemLength: number, maxLength: number) => `${itemLength} / ${maxLength}`;
+
   @State() counterString: string | null | undefined;
 
+  @Watch('counterFormatter')
+  counterFormatterChanged() {
+    if (this.counter) {
```
Comment (liamdebeasi) [inline core/src/components/item/item.tsx:149]: This is redundant. The `updateCounterOutput` call already have an `this.counter` check

File: core/src/components/item/readme.md [hunk 2]
```diff
@@ -1509,6 +1582,48 @@ export class ItemExample {
...
+import { Component, h } from '@stencil/core';
+
+@Component({
+  tag: 'item-example',
+  styleUrl: 'item-example.css'
+})
+export class ItemExample {
+  render() {
+    return [
+      <ion-item counter="true">
```
Comment (liamdebeasi) [inline core/src/components/item/readme.md:1597]:
```suggestion
      <ion-item counter={true}>
```

File: core/src/components/item/readme.md [hunk 3]
```diff
@@ -1509,6 +1582,48 @@ export class ItemExample {
...
+import { Component, h } from '@stencil/core';
+
+@Component({
+  tag: 'item-example',
+  styleUrl: 'item-example.css'
+})
+export class ItemExample {
+  render() {
+    return [
+      <ion-item counter="true" counterFormatter={(inputLength, maxLength) => `${maxLength - inputLength} characters remaining`}>
```
Comment (liamdebeasi) [inline core/src/components/item/readme.md:1618]:
```suggestion
      <ion-item counter={true} counterFormatter={(inputLength, maxLength) => `${maxLength - inputLength} characters remaining`}>
```

**Review by liamdebeasi — CHANGES_REQUESTED** [review]

File: core/src/components/item/usage/vue.md [hunk 4]
```diff
@@ -410,3 +410,38 @@ export default defineComponent({
 });
 </script>
 ```
+
+### Item Counter
+
+```html
+<template>
+  <ion-item :counter="true" :counter-formatter="counterFormatter">
```
Comment (liamdebeasi) [inline core/src/components/item/usage/vue.md:418]: Should these two examples be reversed? The second one says "item counter formatter" but does not use the formatter.

Comment (liamdebeasi) [reply core/src/components/item/usage/vue.md:418]: Looks like this also happens elsewhere

**Review by liamdebeasi — APPROVED** [review]
One non-blocking question: Should `counterFormatter` default to `undefined`? That is what we do with the `isDateEnabled` property. I don't think if it matters here. The only benefit I can think of is a simpler stack trace and more consistency with other APIs.

Otherwise, good to go once the docs changes are updated.

**Pull request merged**